Add fuller steps to pre-release doc (#2056)
alphagov/govuk-frontend: docs/releasing/publishing-a-pre-release.md

# Publish pre-release of GOV.UK Frontend

Use pre-releases when you:
- [work on developing a component or pattern](https://design-system.service.gov.uk/community/develop-a-component-or-pattern/) for the GOV.UK Design System
- want to trial an experimental feature (guidance on trialling experimental features is in development)

> :warning:Your projects should never depend on a pre-released GOV.UK Frontend package. This is because someone could remove the GitHub branch containing the pre-release package at any time. For this reason, never use a pre-released package in a production setting.

## What happens when you pre-release GOV.UK Frontend

When you pre-release GOV.UK Frontend, this creates a GitHub branch. This branch contains the GOV.UK Frontend `/package` directory with your trial changes.

> Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.google.com/url?q=https://www.npmjs.com/package/govuk-frontend&sa=D&ust=1607000302220000&usg=AOvVaw3e1qPUl6v9QAj2LVro13kO). No changes publish to the GOV.UK Frontend npm package as part of this process.

Reviewer: Not sure this last sentence quite makes sense? Perhaps: `No changes are published to the GOV.UK Frontend npm package as part of this process`?

Author: ```suggestion
Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.google.com/url?q=https://www.npmjs.com/package/govuk-frontend&sa=D&ust=1607000302220000&usg=AOvVaw3e1qPUl6v9QAj2LVro13kO). No changes are published to the GOV.UK Frontend npm package as part of this process.
```

> Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.google.com/url?q=https://www.npmjs.com/package/govuk-frontend&sa=D&ust=1607000302220000&usg=AOvVaw3e1qPUl6v9QAj2LVro13kO). No changes are published to the GOV.UK Frontend npm package as part of this process.

Reviewer 2: Does this link URL need tidying up? I think we can link directly to https://www.npmjs.com/package/govuk-frontend.

Author: ```suggestion
Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.npmjs.com/package/govuk-frontend). No changes are published to the GOV.UK Frontend npm package as part of this process.
```

## Publish a pre-release

> 1. Run `git checkout -b BRANCH-FOR-PRE-RELEASE`.

Reviewer 2: I think we've usually used the syntax `<BRANCH-FOR-PRE-RELEASE>` but I might be wrong.

Author: @hannalaakso Mark is drafting some guidance on how to write placeholders. This will say to display placeholders using all caps and hyphens, but not within arrows or square brackets. The reason for this is a concern that some users might include arrows / square brackets within their commands.

2. Run `git pull` to pull the latest changes.

3. Run `nvm use` to make sure you’re using the right version of Node.js and npm.

4. Run `npm install` to make sure you have the latest dependencies installed.

> 5. Run `npm run pre-release` to create and push a new branch, and commit the changes you've made. This process may take a few moments.

Reviewer 2: I think the order here might be the wrong way around. 
> Make a code change and commit the changes you've made

should be step 5. 

> Run `npm run pre-release` to create and push a new branch, and commit the changes you've made. This process may take a few moments.

should be step 6.

Author: ```suggestion
5. Make a code change and commit the changes you've made.
6. Run `npm run pre-release` to create and push a new branch. This process may take a few moments.
```

> 6. Copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/#support) and we'll help you.

Author: ```suggestion
6. Copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/#support) and we'll help you.
```
```suggestion
7. Copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/#support) and we'll help you.
```

> 7. Optionally, to update another project to use the pre-release, navigate to the project in Terminal and run the command there.

Reviewer 2: This is just a suggestion but it might be good to add something like the following at the end of step 7:
> Running this command will make the projects to point to the pre-release branch in their package.json, instead of to the published [GOV.UK Frontend npm package] so that you can preview your trial changes to GOV.UK Frontend.

Author: ```suggestion
7. Optionally, to update another project to use the pre-release, navigate to the project in Terminal and run the command there. Running this command makes the project point to the pre-release branch in your package.json, instead of to the published [GOV.UK Frontend npm package](https://www.npmjs.com/package/govuk-frontend). This enables you to preview your trial changes to GOV.UK Frontend.```

Author: @hannalaakso I added this content to the end of the second last step when I edited the whole file in GitHub. Addition visible here: https://github.com/alphagov/govuk-frontend/pull/2056/files

# Publish pre-release of GOV.UK Frontend

> Before you use a pre-release, you need to make changes to GOV.UK Frontend. Then follow these instructions.

Author: @hannalaakso @36degrees Is there any content we could link to from the text "make changes to GOV.UK Frontend"?

Reviewer 2: @EoinShaughnessy Maybe something like:
> Before you publish a pre-release, you will need to have [committed](https://github.com/alphagov/design-system-team-docs/blob/master/development/commit-messages.md) a code change to GOV.UK Frontend.

But the resource I linked to but might not be the best one as it's about writing a commit message rather than the mechanics of committing a change in git - we don't have guidance for the latter as it's beyond our team docs and depends what tool you're using for git etc. I think "committing a code change" could be enough in this instance.

Reviewer 2: I would use the word "publish" instead of "use" here as it's more accurate.

Author: @hannalaakso Great suggestion! Makes sense to me. Will run by reviewer.

> Before you use a pre-release, you need to make changes to GOV.UK Frontend. Then follow these instructions.

Author: ```suggestion
This pre-release guidance is aimed at Design System team members. If you're an external contributor who needs to create a pre-release, please [contact the Design System team](https://design-system.service.gov.uk/get-in-touch/) and we'll do it for you.

Before you use a pre-release, you need to make changes to GOV.UK Frontend. Then follow these instructions.
```

Author: @hannalaakso ^ Suggested text if we're encouraging external contributors to ask us instead of creating the pre-release themselves.

Reviewer 2: @EoinShaughnessy Yes I think this works 👍

> :warning:Your projects should never depend on a pre-released GOV.UK Frontend package. This is because someone could remove the GitHub branch containing the pre-release package at any time. For this reason, never use a pre-released package in a production setting.

## What happens when you pre-release GOV.UK Frontend

When you pre-release GOV.UK Frontend, this creates a GitHub branch. This branch contains the GOV.UK Frontend `/package` directory with your trial changes.

Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.npmjs.com/package/govuk-frontend). No changes are published to the GOV.UK Frontend npm package as part of this process.

## Publish a pre-release

1. Run `git checkout -b BRANCH-NAME` to switch to the branch you want to pre-release.

2. Run `nvm use` to make sure you’re using the right version of Node.js and npm.

3. Run `npm install` to make sure you have the latest dependencies installed.

4. Run `npm run pre-release` to create and push a new branch that contains your changes. This process may take a few moments and will display a `Success!` message.

## Preview your changes

> 1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, https://design-system.service.gov.uk/get-in-touch/) and we'll help you.

Author: ```suggestion
1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/get-in-touch/) and we'll help you.
```

> 1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/get-in-touch/) and we'll help you.

Author: @hannalaakso If this content is going to be aimed at Design System members, can we remove the sentence here about contacting the Design System team?

Reviewer 2: @EoinShaughnessy I think that's a good call 👍

> 1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/get-in-touch/) and we'll help you.

Author: ```suggestion
1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message.
```

# Publish pre-release of GOV.UK Frontend

This pre-release guidance is aimed at Design System team members. If you're an external contributor who needs to create a pre-release, please [contact the Design System team](https://design-system.service.gov.uk/get-in-touch/) and we'll do it for you.

> Before you use a pre-release, you need to make changes to GOV.UK Frontend. Then follow these instructions.

Author: ```suggestion
Before you publish a pre-release, you need to have committed a code change to GOV.UK Frontend. Then follow these instructions.
```

Author: @hannalaakso The pre-i reviewer liked the change you suggested. I've now sent the content for 2i.

> :warning:Your projects should never depend on a pre-released GOV.UK Frontend package. This is because someone could remove the GitHub branch containing the pre-release package at any time. For this reason, never use a pre-released package in a production setting.

## What happens when you pre-release GOV.UK Frontend

When you pre-release GOV.UK Frontend, this creates a GitHub branch. This branch contains the GOV.UK Frontend `/package` directory with your trial changes.

Projects can point to this branch in their package.json, instead of to the published [GOV.UK Frontend npm package](https://www.npmjs.com/package/govuk-frontend). No changes are published to the GOV.UK Frontend npm package as part of this process.

## Publish a pre-release

> 1. Run `git checkout -b BRANCH-FOR-PRE-RELEASE` to create a new local branch of the GOV.UK Frontend repo.

Reviewer 3: The scope of this has changed from 'publishing a pre-release (of some work that's already been done)' to 'making changes to frontend and then publishing them as a pre-release'.

I can see the benefit of doing so, but I'm wondering if either the title needs to change or we need some sort of introductory text. I think there's a good chance that someone comes in to this document having already made the changes on a branch, and if they follow these instructions they're going to create a branch that they don't need and potentially get quite confused when they reach step 5.

Author: I don't think the current title (Publish pre-release of GOV.UK Frontend) is as descriptive as it could be. It tells users what they'll do, but not why they should do it.

Could 'Preview changes with pre-releases' work as a title? @36degrees @hannalaakso

Reviewer 3: The more I think about it, the less sure I am about this change.

As it stands, I don't think step 5 provides enough detail for the user to actually be able to do it. Summarising it as 'make the changes' glosses over a huge amount of information that most users are going to need – for example, where to actually make the changes, how to know whether the changes have worked (for example how to run the review app) and how to deal with issues that come up (for example, the linter raising errors)

If we actually added that detail, the scope of this document would need to change entirely in order to accommodate it. Without the detail, it's incomplete and forces the user to break out of the steps and go and read other documentation.

With that in mind, could we consider scoping this tightly around the steps to publish a pre-release (assuming you have already made the changes you want to make) and to instead frame the guidance so that users understand how it fits into the wider context and exactly what we expect them to have already done?

Author: @36degrees Not sure I follow — can we have a quick call tomorrow to flesh out what that scoping would look like?

Reviewer 3: Sure thing. It'd be good to invite @hannalaakso as well.

Author: Whoops, forgot to resolve this. Will do that now I've sent the updated content for review.

Author: Whoops, forgot to resolve this. Will do that now I've sent the updated content for review.

> 2. Run `git pull` to pull the latest changes.

Reviewer 3: What is it we're trying to do by running `git pull` at this point? We've only just created this branch locally, so it's not going to be tracking a remote branch yet.

If I try this locally, I get:

```
There is no tracking information for the current branch.
Please specify which branch you want to rebase against.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> my-branch
```

Would it make more sense for the first steps to be checking out the `master` branch and running `git pull` there to ensure they have the latest changes to `master` _before_ then creating a branch to do the work on?

Author: ```suggestion
2. Run `git checkout master` to switch to the master branch and check whether your local branch is up to date.
```

Author: I'll resolve this thread - since these comments were made, we've discussed how to update the doc without including this step.

3. Run `nvm use` to make sure you’re using the right version of Node.js and npm.

4. Run `npm install` to make sure you have the latest dependencies installed.

5. Make a code change and commit the changes you've made.

6. Run `npm run pre-release` to create and push a new branch that contains your changes. This process may take a few moments and will display a `Success!` message.

## Preview your changes

> 1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, [contact the Design System team](https://design-system.service.gov.uk/#support) and we'll help you.

Reviewer 3: I wonder if https://design-system.service.gov.uk/get-in-touch/ might be a better place to link to for support?

Author: ```suggestion
1. If you need to update an existing project to use the pre-release, copy the command that displays after the `Success!`message. If the `Success!` message does not display, https://design-system.service.gov.uk/get-in-touch/) and we'll help you.
```
